### PR TITLE
Scope the projection selection properly when using the mutation conventions.

### LINKED ISSUE #5300: UseProjection on mutation with mutation conventions
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

as per the hotchocolate [#general slack channel thread](https://hotchocolategraphql.slack.com/archives/CD9TNKT8T/p1660314283441189), we are using hot chocolate 12.7 with the following mutation (actual entity and property names are replaced)
```
mutation CreateEntity {
  createEntity (input: {
    foo: 1,
    bar: "aaa",
    childId: 5
  }){
    entity {
      id,
      foo,
      bar,
      childId,
      child {
        id,
        name
      }
    }
  }
}
```
The expected response should look like below (note that both the childId is returned and the child property itself is expanded)
```
{
  "data": {
    "createEntity": {
      "entity": {
        "id": 1234,
        "foo": 1,
        "bar": "aaa",
        "childId": 5,
        "child": {
          "id": 5,
          "name": "foobar"
        }
      }
    }
  }
}
```
The actual response however looks like this:
```
{
  "data": {
    "createEntity": {
      "entity": {
        "id": 1234,
        "foo": 1,
        "bar": "aaa",
        "childId": 5,
        "child": null
      }
    }
  }
}
```
The mutation method looks like this:
```
	[HotChocolate.Data.UseFirstOrDefault]
	public async Task<IQueryable<Entity>> CreateEntity(int foo,
                string bar,
                int childId,
                [Service] SomeService service,
                [Service] DataContext dataContext,
                CancellationToken cancellationToken = default)
	{
		int id = await service.CreateEntity(foo,
				bar,
                                childId,
				cancellationToken);

		return dataContext.Entities.Where(cd => cd.Id == id);
	}
```

We have tried to apply the UseProjection attribute as well, thinking that that may be the reason why we are not getting the expanded navigation property `child` but hotchocolate then throws the following exception:
```
"Unable to cast object of type 'System.Linq.Expressions.Expression1`1[System.Func`2[System.Object,System.Object]]' to type 'System.Linq.Expressions.Expression`1[System.Func`2[Entity,Entity]]'.",
```
```
at HotChocolate.Data.Projections.Expressions.QueryableProjectionScopeExtensions.Project[T](QueryableProjectionScope scope)
   at HotChocolate.Data.Projections.Expressions.QueryableProjectionContextExtensions.Project[T](QueryableProjectionContext context)
   at HotChocolate.Data.Projections.Expressions.QueryableProjectionProvider.<>c__5`1.<CreateApplicatorAsync>b__5_0(IResolverContext context, Object input)
   at HotChocolate.Data.Projections.Expressions.QueryableProjectionProvider.<>c__DisplayClass4_0`1.<<CreateExecutor>g__ExecuteAsync|1>d.MoveNext()
--- End of stack trace from previous location ---
   at HotChocolate.Data.Projections.FirstOrDefaultMiddleware`1.InvokeAsync(IMiddlewareContext context)
   at HotChocolate.Utilities.MiddlewareCompiler`1.ExpressionHelper.AwaitTaskHelper(Task task)
   at HotChocolate.Types.MutationConventionMiddleware.InvokeAsync(IMiddlewareContext context)
   at HotChocolate.Types.ErrorMiddleware.InvokeAsync(IMiddlewareContext context)
   at HotChocolate.AspNetCore.Authorization.AuthorizeMiddleware.InvokeAsync(IDirectiveContext context)
   at HotChocolate.Utilities.MiddlewareCompiler`1.ExpressionHelper.AwaitTaskHelper(Task task)
   at HotChocolate.Execution.Processing.Tasks.ResolverTask.ExecuteResolverPipelineAsync(CancellationToken cancellationToken)
   at HotChocolate.Execution.Processing.Tasks.ResolverTask.TryExecuteAsync(CancellationToken cancellationToken)"
```

@michaelstaib is suggesting that this may be connected to the rewriting of the schema when mutation conventions are applied...

### Steps to reproduce

1. have an entity framework core entity with a navigation property
2. create a graphql mutation (with mutation conventions applied) that exposes a method to persist an instance of this entity and return the newly created entity loaded from the data context (as IQueryable<T> with UseFirstOrDefault applied)
3. call the mutation method and request the result to return the entity with the navigation property expanded
4. hotchocolate returns the entity, however the navigation property is not expanded, it is null (as per above code snippets)

### Relevant log output

_No response_

### Additional Context?

_No response_

### Product

Hot Chocolate

### Version

12.7.0

### PR DESCRIPTION
**Pull request by hahn-kev**
- added some context data to get the MutationPayload data field name
- used that field name to fetch the proper selection to pass down to the projection executor
- wrote some projection integration tests to verify this works as expected

Closes #5300

### CODE REVIEW

**Comment by michaelstaib** [comment]
This looks already good ... thanks for contributing... I will have a look at it later tonight.

File: src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs [hunk 1]
```diff
@@ -205,10 +206,28 @@ private static FieldMiddleware CreateMiddleware<TEntity>(IProjectionConvention c
...
             }
 
+            if (context.Operation.Definition.Operation == OperationType.Mutation
+                && context.Selection.Field.Type.NullableType() is ObjectType mutationPayloadType
+                && mutationPayloadType.ContextData.GetValueOrDefault(MutationConventionDataField, null) is string dataFieldName)
+            {
+                var dataField = mutationPayloadType.Fields[dataFieldName];
+
+                var selection = UnwrapMutationPayloadSelect(context, dataField);
+                context = new MiddlewareContextProxy(context, selection, dataField.DeclaringType);
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs:216]:
```suggestion
                var dataField = mutationPayloadType.Fields[dataFieldName];
                var selection = UnwrapMutationPayloadSelect(context, dataField);
                context = new MiddlewareContextProxy(context, selection, dataField.DeclaringType);
```

File: src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs [hunk 2]
```diff
@@ -205,10 +206,27 @@ private static FieldMiddleware CreateMiddleware<TEntity>(IProjectionConvention c
                 context = new MiddlewareContextProxy(context, selection, objectType);
             }
 
+            if (context.Operation.Definition.Operation == OperationType.Mutation
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs:209]: should this be an else if or do we want to run this always?

File: src/HotChocolate/Data/test/Data.Projections.Tests/IntegrationTests.cs [hunk 3]
```diff
@@ -289,6 +339,24 @@ public IQueryable<Foo> Foos
         => new Foo[] { new() { Bar = "A" }, new() { Bar = "B" } }.AsQueryable();
 }
 
+public class Mutation
+{
+    [UseMutationConvention]
+    [UseProjection]
+    public IQueryable<Foo> Modify()
```
Comment (PascalSenn) [inline src/HotChocolate/Data/test/Data.Projections.Tests/IntegrationTests.cs:346]: Can we have one test here that shows that it also works when there is an error? I mean, that it not just blows up completely

File: src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs [hunk 4]
```diff
@@ -205,10 +206,27 @@ private static FieldMiddleware CreateMiddleware<TEntity>(IProjectionConvention c
...
+                && context.Selection.Field.Type.NullableType() is ObjectType mutationPayloadType
```
Comment (PascalSenn) [inline src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs:210]: @michaelstaib  would this be clearer?
```suggestion
                && context.Selection.Field.Type.NamedType() is ObjectType mutationPayloadType
```

**Review by PascalSenn — COMMENTED** [review]
Looks really good!

File: src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs (hunk 2 above)
Comment (hahn-kev) [reply src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs:209]: that's a really good question. I think a mutation can return a relay resolver, I'll see what order it makes sense to apply those. That might be a difficult edge case to cover however.

File: src/HotChocolate/Data/test/Data.Projections.Tests/IntegrationTests.cs (hunk 3 above)
Comment (hahn-kev) [reply src/HotChocolate/Data/test/Data.Projections.Tests/IntegrationTests.cs:407]: Done

File: src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs (hunk 2 above)
Comment (hahn-kev) [reply src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs]: looks like an `else if` will work fine here. I pushed some tests with node resolvers in a mutation and it worked just fine.

File: src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs [hunk 5]
```diff
@@ -204,11 +205,29 @@ private static FieldMiddleware CreateMiddleware<TEntity>(IProjectionConvention c
                 var selection = CreateProxySelection(context.Selection, fieldProxy);
                 context = new MiddlewareContextProxy(context, selection, objectType);
             }
-
+            //for use case when projection is used with Mutation Conventions
+            else if (context.Operation.Definition.Operation == OperationType.Mutation
+                     && context.Selection.Field.Type.NamedType() is ObjectType mutationPayloadType
+                     && mutationPayloadType.ContextData.GetValueOrDefault(MutationConventionDataField, null) is string
+                         dataFieldName)
```
Comment (michaelstaib) [inline src/HotChocolate/Data/src/Data/Projections/Extensions/ProjectionObjectFieldDescriptorExtensions.cs:212]:
```suggestion
            else if (context.Operation.Definition.Operation == OperationType.Mutation && 
                context.Selection.Field.Type.NamedType() is ObjectType mutationPayloadType && 
                mutationPayloadType.ContextData.GetValueOrDefault(MutationConventionDataField, null) is string dataFieldName)
```

**Review by michaelstaib — APPROVED** [review]

**Comment by michaelstaib** [comment]
@hahn-kev I did a couple more refinements ... but I think its good now.

**Comment by michaelstaib** [comment]
I created a new preview of 13.5 ... you can use it with the newest preview.

**Comment by hahn-kev** [comment]
Tested it out and it works exactly how I expected. Looking forward to the full release. I'll try and use it and let you know if we find issues.
